src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -5922,12 +5938,22 @@ private void ParseTypeArgumentList(out SyntaxToken open, SeparatedSyntaxListBuil
             // remaining types & commas
             while (true)
             {
-                if (this.CurrentToken.Kind == SyntaxKind.GreaterThanToken)
+                // We prefer parsing as if there was a missing > over parsing as a tuple type for error recovery
+                if (tokenKindBreaksArgumentList(this.CurrentToken.Kind) || this.CurrentToken.Kind is SyntaxKind.GreaterThanToken)
                 {
                     break;
                 }
                 else if (this.CurrentToken.Kind == SyntaxKind.CommaToken || this.IsPossibleType())
```

Comment:
as per above, i feel like it would make sense to break this into separate chunks.  For example, the "this.currenttoken.kind is identifeir" can't ever be true for the "this.currenttoken.kind is comma" check above.  

this would clarify things a lot.